### DISCUSSION TITLE
Simplify E-Cash Mint Server Module Consensus Settings

### DISCUSSION THREAD

**Discussion by tvolk131** [fedimint/fedimint#3327] [Fedimint Modules]
I'm looking for feedback on a design proposal surrounding the E-Cash mint server module. The consensus settings for this module contains a field `mint_amounts` which allows for arbitrary denomination amounts to be issued:

https://github.com/fedimint/fedimint/blob/9d871a719056415503d92e7d73e05c53d4b145f0/modules/fedimint-mint-common/src/config.rs#L20-L22

Currently we only ever use base-2 amounts, and we do this by generating the amounts and setting mint_amounts to the generated values:

https://github.com/fedimint/fedimint/blob/9d871a719056415503d92e7d73e05c53d4b145f0/fedimintd/src/lib.rs#L55-L60

https://github.com/fedimint/fedimint/blob/9d871a719056415503d92e7d73e05c53d4b145f0/modules/fedimint-mint-common/src/config.rs#L29-L34

https://github.com/fedimint/fedimint/blob/9d871a719056415503d92e7d73e05c53d4b145f0/modules/fedimint-mint-server/src/lib.rs#L789-L791

In the first two examples above, we can also see that the values are being generated using a specified maximum denomination value.

While it'd be nice to retain the ability to use other bases for denominations, to my knowledge there is no reason to retain the ability to issue arbitrary note amounts with no shared denomination base. I propose we replace the `mint_amounts` field with fields that can be used to compute the denomination amounts. We could use, say, `denomination_base` and `max_denomination_size` which would be enough information to determine the existing `mint_amounts` and therefore replace that field. We could even get rid of `max_denomination_size` and generate values all the way up to 21,000,000 coins (which, if I'm doing my math correctly, would require 35 base-2 denominations to reach instead of 14 to reach the existing 10 coin limit). We could add a method to the struct to generate the denomination amounts (i.e. consolidate the calls to `Tiered::gen_denominations()` in the links above). I think this would make mint configuration more declarative and easy to grok.

Does this make sense? Happy to hear feedback, thanks for taking the time to read!

**Comment by wbobeirne** [comment]
> While it'd be nice to retain the ability to use other bases for denominations, to my knowledge there is no reason to retain the ability to issue arbitrary note amounts with no shared denomination base.

One use-case we've arrived at is being able to make very small QR codes for offline sends. If you were going to make some printable ecash notes that you wanted to fit in a single QR code, you could pre-specify nice round denominations (e.g. 10,000 sats) so that they would be easily scannable. Otherwise you can end up with ecash strings that are 1000+ characters long and can't easily  fit.

**Reply by wbobeirne** [reply]
However that's somewhat of a niche use case, and there may be better ways to do this than bespoke denominations that require consensus. Definitely open to alternatives for that!

**Reply by tvolk131** [reply]
Funny, I actually just talked to one of the folks at [Noteworthy](https://noteworthy.ag/) today at their table in Pacific Bitcoin about the potential of printing physical paper Fedi ECash bills! They seemed timidly interested and asked for my contact info.

How many notes could easily fit in a single QR code? If it's at least 5, then the easiest solution may be to run a federation using base-10 denominations. That way you could make 20-sat bills using 2 10-sat ECash notes, or 50-sat bills using 5 10-sat ECash notes. If you can fit 7 in a QR code, you can make 25-sat bills using 2 10-sat and 5 1-sat ECash notes. And since it's base 10, you could obviously slide the decimal point around.

**Reply by wbobeirne** [reply]
> How many notes could easily fit in a single QR code?

Unfortunately it's quite the opposite 😅 QR codes can contain a maximum of ~180 characters before they start to become unreadable on some devices. Even a single ecash note requires a minimum of 3 QR codes to support being easily scanned a wide range of capable devices (Split across multiple QR codes using a technique like [qrloop](https://github.com/gre/qrloop))

From there, each note adds about 1.5 QR codes (rounding up) per note denomination. Even the best combinations of round-ish numbers that are optimized for minimal note usage in base 2 denominations tend to be 4-6 notes, so like 9-10 QR codes. And to account for the imperfection of physical printing and wear and tear, you'd want to set a decently high error correction level (at least M - 15% correction, maybe higher) which further pads them in size.

It may just be that for physical notes, we would want to consider an alternative compressed way of exchanging ecash.

**Reply by dpc** [reply]
`1048576` (2^20 msats) is nice and round to me. :D

**Reply by dpc** [reply]
I created https://github.com/fedimint/fedimint/issues/3342 to see if there's anything we could do to make these notes smaller. Got blocked on `just mprocs` not working, ATM.

**Reply by humansinstitute** [reply]
A potentially simpler alternative aproach here could be to lock the eCash into a hash lock contract and then use the QR Code to communicate the pre-image to a new user?

**Comment by dpc** [comment]
One reason to do it: https://github.com/fedimint/fedimint/issues/3342#issuecomment-1755997813

**Reply by elsirion** [reply]
This isn't a good reason imo as we can just use the denomination index, so even if we had 1,2,5,10,20,50,… denominations we could encode 5 as 2 because it's the third denomination.

**Comment by elsirion** [comment]
I think currently the client wouldn't work with denominations not in the form `base^n` anyway, we made that decision a while ago. Idk if having some special denominations in addition to the `base^n` ones works with note selection, feels like it should.

I'd be ok with only allowing `base^n` ones as long as we don't get locked in on `base=2` because that's a privacy trade-off.

**Reply by tvolk131** [reply]
> I propose we replace the mint_amounts field with fields that can be used to compute the denomination amounts. We could use, say, `denomination_base` and `max_denomination_size` which would be enough information to determine the existing `mint_amounts` and therefore replace that field.

So assuming we can find a way around the offline QR compaction problem, does this solution seem reasonable?

> We could even get rid of `max_denomination_size` and generate values all the way up to 21,000,000 coins

What are your thoughts on this? I don't see a strong case for capping our denomination sizes since the cost of generating denominations all the way up to Bitcoin's supply cap is low and imo far lower than the cost of code complexity to maintain an extra configurable parameter.

**Reply by dpc** [reply]
I'm also not sure of any use for `max_denomination_size`. Was it supposed to force people into relatively smaller denomination to increase anonimity set for higher amounts? If someone pegs-in 1000k BTC, then after some time withdraws 1000k BTC, does the fact that it was split into 1 BTC notes, improve much for anyone?

On the other hand we did have real problems when testing when we peg-ed in 100BTC (regtest or signet), and then were unable to complete peg-in because maximum denonmination was like 0.1BTC or something and a transaction would have 1k or 10k inputs or something like that.

It seems like an unnecessary foot-gun. Not sure if going up to 21M is necessary, but I'd just make it 1M BTC, and not let users touch it.

**Reply by elsirion** [reply]
Yes, let's get rid of that maximum, shouldn't make key gen that much slower.

**Comment by dpc** [comment]
@tvolk131 I guess there's no opposition and even some benefits, and we figured out some details too, so feel free to work on it.

**Reply by tvolk131** [reply]
Created an issue that consolidates this conversation! #3383

Thanks everyone for the input!
